src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Fetches all HA-flows.
+     */
+    public List<HaFlowResponse> getAllHaFlows() {
+        return transactionManager.doInTransaction(getReadOperationRetryPolicy(), haFlowRepository::findAll)
+                .stream()
+                .map(haFlow -> HaFlowMapper.INSTANCE.toHaFlowDto(haFlow, flowRepository, haFlowRepository))
+                .map(HaFlowResponse::new)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets HA-flow by ID.
+     */
+    public HaFlowResponse getHaFlow(@NonNull String haFlowId) throws FlowNotFoundException {
+        return transactionManager.doInTransaction(getReadOperationRetryPolicy(), () ->
+                        haFlowRepository.findById(haFlowId))
+                .map(haFlow -> HaFlowMapper.INSTANCE.toHaFlowDto(haFlow, flowRepository, haFlowRepository))
+                .map(HaFlowResponse::new)
+                .orElseThrow(() -> new FlowNotFoundException(haFlowId));
+    }
+
+    /**
+     * Gets HA-flow paths by ID.
+     */
+    public HaFlowPathsResponse getHaFlowPaths(@NonNull String haFlowId) throws FlowNotFoundException {
+        return transactionManager.doInTransaction(getReadOperationRetryPolicy(), () -> {
+            HaFlow haFlow = haFlowRepository.findById(haFlowId).orElseThrow(() -> new FlowNotFoundException(haFlowId));
+
+            List<FlowPathDto> subFlowPathDtos = new ArrayList<>();
+            Map<String, FlowPath> forwardPaths = new HashMap<>();
+            Map<String, FlowPath> protectedForwardPaths = new HashMap<>();
+            Map<String, FlowPath> reversePaths = new HashMap<>();
+            Map<String, FlowPath> protectedReversePaths = new HashMap<>();
+
+            for (FlowPath flowPath : haFlow.getForwardPath().getSubPaths()) {
+                forwardPaths.put(flowPath.getHaSubFlowId(), flowPath);
+            }
+
+            for (FlowPath flowPath : haFlow.getReversePath().getSubPaths()) {
+                reversePaths.put(flowPath.getHaSubFlowId(), flowPath);
+            }
+
+            if (haFlow.isAllocateProtectedPath()) {
+                for (FlowPath flowPath : haFlow.getProtectedForwardPath().getSubPaths()) {
+                    protectedForwardPaths.put(flowPath.getHaSubFlowId(), flowPath);
+                }
+                for (FlowPath flowPath : haFlow.getProtectedReversePath().getSubPaths()) {
+                    protectedReversePaths.put(flowPath.getHaSubFlowId(), flowPath);
+                }
+            }
+
+            for (HaSubFlow haSubFlow : haFlow.getHaSubFlows()) {
+                String haSubFlowId = haSubFlow.getHaSubFlowId();
+
+                FlowPathDtoBuilder flowPathDtoBuilder = buildFlowPathDto(haFlow, forwardPaths.get(haSubFlowId),
+                        reversePaths.get(haSubFlowId));
+                flowPathDtoBuilder.protectedPath(buildProtectedPathDto(haFlow, protectedForwardPaths.get(haSubFlowId),
```

Comment:
Am I right that we always include protected path it the response, even if flow has no protected path?

Could you please check if we include empty protected info for common flows and Y flows? if we do not include it - please do the same

Reply:
You are right. An empty protected path is redundant. Will fix it.

Reply:
Fixed